[FEAT/#4] 비대칭 키 Token Provider 구현

hyunw9 opened the pull request:
## Related Issue 🚀
- closed #4 

## Work Description ✏️
- ATK, RTK 객체를 정의하고 TokenProvider 의 구현체인 JwtAuthProvider, JwtRefreshProvider를 구현했습니다.
단위테스트는 Spring Bean이 필요하기 때문에 통합테스트로 진행하였습니다.

## PR Point 📸

### [토큰 객체 구현](https://github.com/sopt-makers/sopt-auth-backend/commit/5a6a9d4a3f532ca75d9b574ee164b042238f8f9d)

이 부분이 이전 코드와 조금 달라졌습니다. 
RefreshToken 과 AccessToken의 역할이 조금 다른 것 같아서, 모델 내부에서 로직을 처리하도록 변경했습니다. 

혹시 수정할 부분이 보이신다면 리뷰 부탁드립니다 ! 

### [토큰 Provider 구현](https://github.com/sopt-makers/sopt-auth-backend/commit/90c821d6a2e5b210cf25b3fe8d83a393b5b166cb)

각 토큰의 Provider를 구현했는데요, ATK,RTK는 초기에 "Bearer " 헤더가 같이 전달됩니다.
이를 인코딩 전에 제거하고, 디코딩 후에 붙여주는 작업은 Provider가 수행하는 것이 적당하다고 생각했습니다. 

따라서 provider는 Token의 생성, prefix의 추가/삭제를 담당합니다. 
TokenUtil을 사용하여, prefix 추가/제거 (검증)을 위임했습니다.

hyunw9 committed:
[FEAT] Jwt key Property 설정 & Config 추가

hyunw9 committed:
[FEAT] 토큰 객체 구현

hyunw9 committed:
[FEAT] TokenProvider 및 Util 추가

sung-silver commented on src/main/java/sopt/makers/authentication/support/common/code/failure/TokenFailure.java:
```diff
@@ -11,6 +11,11 @@
 @Getter
 @RequiredArgsConstructor(access = AccessLevel.PRIVATE)
 public enum TokenFailure implements FailureCode {
+  TOKEN_EXPIRED(HttpStatus.BAD_REQUEST, "토큰이 만료되었습니다."),
```
`p2`

저희 실패코드 AUTH_001 이런식으로 관리하기로하지 않았나요?!
테이블 형태로 만들어 노션에서 관리해야할 것 같습니다!

yummygyudon replied:
@sung-silver 
실패코드의 경우, 저희가 아직 논의를 제대로 하고 컨벤션을 확정짓지 않았고 상세 에러 코드 관리 정책이 존재하지 않기 때문에
아직 사용하기에는 어려울 것이라고 생각됩니당!

hyunw9 replied:
네에, 일단은 구현단계에 집중을 하는걸로 하겠습니다 !

sung-silver commented on src/main/java/sopt/makers/authentication/support/constant/JwtConstant.java:
```diff
@@ -0,0 +1,11 @@
+  public static final String[] ISSUERS = {"operation", "playground"};
+  public static final String ISSUER = "operation";
+
+  public static final long ACCESS_TOKEN_EXPIRATION = 1000L * 60 * 10;
+  public static final long REFRESH_TOKEN_EXPIRATION = 1000L * 60 * 60 * 24 * 7;
+  public static final String TOKEN_HEADER = "Bearer ";
```
`p1`

요부분은 Yml에 들어가야할 내용인 것 같은데 어떻게 생각하시나요..?!

hyunw9 replied:
동의합니다 ..

yummygyudon replied:
동의합니다!

sung-silver commented on src/main/java/sopt/makers/authentication/support/jwt/provider/JwtTokenUtil.java:
```diff
@@ -0,0 +1,17 @@
+  }
+
+  public String addPrefix(String token) {
+    return JwtConstant.TOKEN_HEADER + token;
```
`p4`

static import는 어떠신가요?!

yummygyudon replied:
저도 성은이 의견 동의합니다!

hyunw9 replied:
수정했습니다  !

sung-silver commented on src/main/java/sopt/makers/authentication/support/jwt/token/JwtAccessToken.java:
```diff
@@ -0,0 +1,107 @@
+
+  private void validateExpiration() {
+    Date expiration = Date.from(this.jwt.getExpiresAt());
+    if (expiration == null || expiration.before(new Date())) {
```
`p3`
연산자로 묶여있어서 설명적 변수로 분리하면 좋을 것 같습니다!

hyunw9 replied:
넵넵 ! 확인했습니다

sung-silver commented on src/test/java/sopt/makers/authentication/jwt/JwtRefreshTokenTest.java:
```diff
@@ -0,0 +1,90 @@
+  @Test
+  @DisplayName("RefreshToken 생성")
+  public void generate_jwt_refresh_token() {
+    // When
+    String refreshToken = jwtAuthRefreshTokenProvider.generate("");
+
+    // Then
+    System.out.println("RefreshToken: [" + refreshToken + "]");
```
`p4`
mocking으로 정해진 문자열을 반환하게 만들어 그 문자열과 같음을 검증하는 로직이 들어가는게 좀 더 자연스럽지 않을까요?
Print문을 통해 테스트를 검증하는게 조금 어색하다는 생각이 들었습니다!

hyunw9 replied:
넵 Assertion으로 수정하겠습니다 !

sung-silver reviewed (changes requested):
퇴근하고 다시 한번 더 보겠습니다!

yummygyudon commented on src/main/java/sopt/makers/authentication/support/jwt/provider/JwtAuthAccessTokenProvider.java:
```diff
@@ -0,0 +1,35 @@
+  @Override
+  public CustomAuthentication parse(String requestToken) throws IOException {
+    String token = tokenUtil.extract(requestToken);
+    JwtAccessToken jwtAccessToken = JwtAccessToken.fromStringToken(token, jwtDecoder);
```
`P2`

Jwt 관심사의 Util 역할을 하는 객체(`JwtEncoder`, `JwtDecoder`)가 파라미터로 주입되도록 한 이유가 뭘까요?!
만약 encode나 decode가 필요하다면 해당 기능을 수행 시킨 후에 String만 주입해도 가능할 것이라고 생각이 드는데요!

yummygyudon replied:
아래 RefreshToken에서는 String만 들어가는 구조인 것 같아서
통일성 측면에서도 조금 어긋난다고 생각하구요!

yummygyudon commented on src/main/java/sopt/makers/authentication/support/jwt/token/JwtAccessToken.java:
```diff
@@ -0,0 +1,107 @@
+            .map(GrantedAuthority::getAuthority)
+            .collect(Collectors.toUnmodifiableList());
+    JwtClaimsSet claimsSet = generateClaimSet(subject, issuer, now, expiration, roles);
+    Jwt jwt = jwtEncoder.encode(JwtEncoderParameters.from(claimsSet));
+    String token = jwtEncoder.encode(JwtEncoderParameters.from(claimsSet)).getTokenValue();
```
`P2`

동일한 데이터에 대해 "감싼 것"과 "안 감싼 것"의 차이로 중복 데이터 저장이 발생하고 있는 것 같습니다.
이렇게 구현하신 이유가 궁금해요!

`Jwt`를 생성자로 주입 받고 Getter 롬복을 사용하지 않고
별도 `getValue` 메서드를 구현하는 것은 어떤지 의견이 궁금합니다.
(객체 이름이 이미`~~Token`이기 때문에 `getTokenValue`와 같은 불필요한 상세는 필요없다고 생각합니다)


```java
private final Jwt jwt;
public JwtAccessToken(final Jwt jwt) {
    ...
}

...
public String getValue() {
    return this.jwt.getTokenValue();
}
```

--- 
추가적으로 public static 생성자 팩토리 메서드가 있음에도 생성자를 public으로 열어둔 이유가 궁금합니다.

yummygyudon commented on src/main/java/sopt/makers/authentication/support/jwt/token/JwtAccessToken.java:
```diff
@@ -0,0 +1,107 @@
+  private void validateIssuer() {
+    String issuer = jwt.getClaim(JwtClaimNames.ISS);
+
+    Optional.ofNullable(issuer)
+        .filter(i -> Arrays.asList(JwtConstant.ISSUERS).contains(i))
+        .orElseThrow(() -> new TokenException(UNSUPPORTED_ISSUER));
```
`P2`

Optional 라이브러리와 stream 방식을 사용하신 이유가 궁금합니다!!
기본적인 if 문으로 지정 가능할 것 같아서요!
```java
if (!Arrays.asList(JwtConstant.ISSUERS).contains(issuer)){
    throw new TokenException(UNSUPPORTED_ISSUER);
}
```

hyunw9 replied:
이 부분은 기능 이관 시 모든 기능을 한번에 이관하지 않기 때문에,
Issuer가 여러명 존재할 수 있다고 생각했습니다 ! (operation, playground) 
(혹시 제가 이해한 부분이 틀렸으면 다시 남겨주시면 감사하겠습니다 ! )

때문에 코드 가독성을 위해 `stream`을 사용하였지만, `Optional`은 불필요한 연산이네요 ! 수정하겠습니다 !

yummygyudon commented on src/main/java/sopt/makers/authentication/support/jwt/token/JwtAccessToken.java:
```diff
@@ -0,0 +1,107 @@
+
+  private void validateSubject() {
+    String subject = jwt.getClaim(JwtClaimNames.SUB);
+    if (subject == null) throw new TokenException(TokenFailure.INVALID_SUBJECT);
```
`P3`

위 `validateIssuer()`에서는 TokenFailure가 static import 되어 사용되는데
해당 메서드에서는 명시하신 이유가 궁금합니다

hyunw9 replied:
핫.. 변경했습니다 !~

yummygyudon reviewed (changes requested):
코드 이관 과정에서 놓친 부분들이 몇 가지 있어보입니다.

Yaml과 Property 구조가 최신 구조가 아닌것 같아요? 확인 부탁드립니다

++ 음 우선 내용이 많아서 1차적으로 여기까지 리뷰하도록 하겠습니다..!! 곧 2차 리뷰 하겠습니다!

hyunw9 committed:
[FEAT] key 등록 방식 변경 & 상수 환경변수화

hyunw9 committed:
[FEAT] AccessToken generate / parsing 방식 수정

Decoder / Encoder를 넘기지 않고 토큰객체로서의 역할(검증,저장) 에 집중하도록 변경했습니다.

또한 기존 JWT + String 멤버 구조에서 JWT 만 보관하도록 변경하였습니다.

hyunw9 committed:
[FEAT] RTK provider & RTK 객체 리팩토링

Decoder / Encoder를 넘기지 않고 토큰객체로서의 역할(검증,저장) 에 집중하도록 변경했습니다.

또한 기존 JWT + String 멤버 구조에서 JWT 만 보관하도록 변경하였습니다.

RTK는 parsing 시점, 즉 조회하는 경우는 ATK가 만료되어서 저장소에서 대응되는 RTK를 조회하는 경우의 수가 존재한다고 생각하여

기존 RTK 토큰 검증 이후 refresh 하도록 구현했습니다.

hyunw9 committed:
[SETTING] .pem 값 인식을 위한 의존성 추가

sung-silver commented on build.gradle.kts:
```diff
@@ -42,8 +42,14 @@ dependencies {
+
 	compileOnly("org.projectlombok:lombok")
 	runtimeOnly("org.postgresql:postgresql")
+	runtimeOnly("com.mysql:mysql-connector-j")
```
`P5`

로컬 환경 및 개발, 운영 환경은 postgresql, 테스트는 h2 데이터베이스로 알고 있는데 mysql을 추가하신 이유가 있을까요?
test 환경도 도커로 postgresql 띄워서 완전히 동일한 환경을 구축하면 좋을 것 같긴하네요 @yummygyudon 어떻게 생각하시는지 궁금합니다!

hyunw9 replied:
이것저것 테스트를 mysql로 하고 있어서 그랬습니다 ㅎㅎ 미처 반영이 안된거같네요 수정하겠습니다 !

sung-silver commented on src/main/java/sopt/makers/authentication/support/common/code/failure/TokenFailure.java:
```diff
@@ -11,6 +11,14 @@
 public enum TokenFailure implements FailureCode {
+  TOKEN_EXPIRED(HttpStatus.BAD_REQUEST, "토큰이 만료되었습니다."),
+  UNSUPPORTED_ISSUER(HttpStatus.BAD_REQUEST, "신뢰할 수 없는 발급자입니다."),
+  INVALID_SUBJECT(HttpStatus.BAD_REQUEST, "주체 정보가 잘못되었습니다."),
+  INVALID_PREFIX(HttpStatus.BAD_REQUEST, "토큰 접두사가 잘못되었습니다."),
+  INVALID_ALGORITHM(HttpStatus.BAD_REQUEST, "알고리즘이 잘못되었습니다."),
+  INVALID_SIGNATURE(HttpStatus.BAD_REQUEST, "서명이 잘못되었습니다."),
+  INVALID_LOCATION(HttpStatus.BAD_REQUEST, "키 파일 위치가 잘못되었습니다."),
```
`P4`

너무 디테일한 정보를 반환하는 것 같은데, 메시지는 저희만 확인할 수 있게하고, 인증관련된 오류는 큰 범주로 '인증에 실패했습니다' 등으로 전달하는건 어떻게 생각하시나요?

hyunw9 replied:
동의합니다 ! 일단 서버 로직 관련 예외는 그대로 처리하고, 구현 완료된 이후 에러코드 정의할때 의견을 맞춰보면 좋을 것 같다고 생각했습니다 !

yummygyudon replied:
@sung-silver 
> 메시지는 저희만 확인할 수 있게하고, 인증관련된 오류는 큰 범주로 '인증에 실패했습니다' 등으로 전달하는 방법

위 방법이 어떤 방법을 말씀하시는 건지 모르겠습니당 헣ㅎ

저는 참고로 제공하는 원인 정보는 최대한 자세할 수록 좋다고 생각하긴합니다!

sung-silver replied:
@yummygyudon 클라이언트가 정상적인 요청을 한다면 위의 예외는 발생하지 않으므로 토큰 관련 예외는 만료된 토큰으로 요청을 보내는 경우를 제외하고는 모두 값을 임의로 변경하여 서버로 요청을 보낸다고 생각했어요!
때문에 어떤 부분에서 잘못되었는지에 대한 정보를 노출하지 말자는 의미로 제안드린 방식입니다!
사람마다 다르겠지만 제가 했던 웹 서비스에서는 클라이언트분들이 status가 401이면 토큰 재발급 요청으로 가도록 로직을 처리해서 message를 사용하지 않기도 했고요!

sung-silver commented on src/main/java/sopt/makers/authentication/support/config/JwtRSAKeyConfiguration.java:
```diff
@@ -0,0 +1,115 @@
+  private final JwtProperty jwtProperty;
+  private final ResourceLoader resourceLoader;
+
+  public JwtRSAKeyConfiguration(JwtProperty jwtProperty, ResourceLoader resourceLoader) {
+    this.jwtProperty = jwtProperty;
+    this.resourceLoader = resourceLoader;
+  }
```
`p4`

@RequireArgsContructor를 사용하지 않으신 이유가 특별히 있을까요?? Property 때문인가..

hyunw9 replied:
크게 두가지 이유 때문이었는데요
1. 명확한 코드 이해를 위해 : RSA key를 불러오고, Property를 주입받아 사용하는 방식에 익숙하지 않다 보니, 명확히 생성자를 통해 생성하는 부분이 심리적으로 도움을 주었습니다(..)
2. Config, Logging, Properties 어노테이션이 많아진다.. ! : 어노테이션이 점점 많아져서 최대한 직접 구현할 수 있는 부분은 어노테이션을 사용하지 말아보자..? 라는 마음가짐에서 사용하지 않은 것 같아요 ! 

비슷한 이유로 Token 객체에서도 사용하지 않고 있었는데, 통일하는게 좋을 것 같네요 ! 

@yummygyudon , @sung-silver

sung-silver commented on src/main/java/sopt/makers/authentication/support/config/JwtRSAKeyConfiguration.java:
```diff
@@ -0,0 +1,115 @@
+      X509EncodedKeySpec keySpec = new X509EncodedKeySpec(publicKeyBytes);
+      KeyFactory keyFactory = KeyFactory.getInstance("RSA");
+      return (RSAPublicKey) keyFactory.generatePublic(keySpec);
+    } catch (IOException e) {
+      throw new TokenException(INVALID_LOCATION);
+    } catch (NoSuchAlgorithmException e) {
+      throw new TokenException(INVALID_ALGORITHM);
+    } catch (InvalidKeySpecException e) {
+      throw new TokenException(INVALID_SUBJECT);
```
`p3`

이부분에서 로깅을 남기고 클라이언트에게 전달되는 예외메시지는 추상적으로 전달하면 좋을 것 같네요!

hyunw9 replied:
아하 위에서 말씀하신 부분이군요 확인했습니다 !

sung-silver commented on src/main/java/sopt/makers/authentication/support/config/JwtRSAKeyConfiguration.java:
```diff
@@ -0,0 +1,115 @@
+  public RSAPublicKey createPublicKeyFromProperty() {
+    Resource resource = resourceLoader.getResource(jwtProperty.secret().rsa().publicKey());
+    try {
+      PemReader pemReader =
+          new PemReader(new StringReader(resource.getContentAsString(StandardCharset.UTF_8)));
+      PemObject pemObject = pemReader.readPemObject();
+      pemReader.close();
+
+      byte[] publicKeyBytes = pemObject.getContent();
+      X509EncodedKeySpec keySpec = new X509EncodedKeySpec(publicKeyBytes);
+      KeyFactory keyFactory = KeyFactory.getInstance("RSA");
+      return (RSAPublicKey) keyFactory.generatePublic(keySpec);
```
`p3`

파일을 읽어 키를 만드는 과정인 것 같은데 가독성이 떨어지는 느낌이 들어서 각 행위들을 나누어 Private 메서드로 분리해보는건 어떨까요? 이미 try 문으로 중첩되어 들어간 구조라 코드가 복잡하게 느껴지는 것 같습니다!

hyunw9 replied:
동의합니다 분리하겠습니다 !

sung-silver commented on src/main/java/sopt/makers/authentication/support/config/JwtRSAKeyConfiguration.java:
```diff
@@ -0,0 +1,115 @@
+  @Bean
+  public RSAPrivateKey createPrivateKeyFromProperty() {
+    Resource resource = resourceLoader.getResource(jwtProperty.secret().rsa().privateKey());
+    try {
```
`p5`

변수 선언과 실행부는 한줄 띄어주면 가독성이 올라갈 것 같습니다!

hyunw9 replied:
아고고 확인입니다

sung-silver commented on src/main/java/sopt/makers/authentication/support/config/JwtRSAKeyConfiguration.java:
```diff
@@ -0,0 +1,115 @@
+  public RSAPrivateKey createPrivateKeyFromProperty() {
+    Resource resource = resourceLoader.getResource(jwtProperty.secret().rsa().privateKey());
+    try {
+      PemReader pemReader =
+          new PemReader(new StringReader(resource.getContentAsString(StandardCharsets.UTF_8)));
+      PemObject pemObject = pemReader.readPemObject();
+      pemReader.close();
+
+      byte[] privateKeyBytes = pemObject.getContent();
+      PKCS8EncodedKeySpec keySpec = new PKCS8EncodedKeySpec(privateKeyBytes);
+
+      KeyFactory keyFactory = KeyFactory.getInstance("RSA");
+      return (RSAPrivateKey) keyFactory.generatePrivate(keySpec);
```
`p3`

위에 리뷰한 내용인데 try문을 사용하여 가독성이 떨어지고 있으니, 변수화될 수 있는 부분을 파라미터로 분리하여 Private 메서드로 추출하면 코드를 재사용할 수 있을 것 같습니다!

sung-silver commented on src/test/java/sopt/makers/authentication/jwt/JwtAccessTokenTest.java:
```diff
@@ -0,0 +1,108 @@
+  void create_jwt_access_token() {
+    // Given
+    CustomAuthentication customAuthentication = new CustomAuthentication("test", "test");
+    // When
```
`P4`

//when 앞에 빈 줄 하나 들어가야할 것 같아요!

hyunw9 replied:
넵 수정했습니다 !

sung-silver commented on src/test/java/sopt/makers/authentication/jwt/JwtRefreshTokenTest.java:
```diff
@@ -0,0 +1,98 @@
+  public void validate_jwt_refresh_token() throws IOException {
+    // Given
+    JwtClaimsSet jwtClaimsSet =
+        JwtClaimsSet.builder().expiresAt(Instant.now().minusSeconds(1)).build();
```
`p4`

여기 불필요한 줄바꿈이 포함된 것 같습니다!

hyunw9 replied:
인텔리제이가 줄을 자동으로 바꿔주는군요. 수정했습니다 !

sung-silver reviewed (changes requested):
코드 가독성 측면에서 리뷰를 많이 한 것 같긴한데 한 번 확인 부탁드립니다!

yummygyudon commented on src/main/java/sopt/makers/authentication/support/jwt/provider/JwtTokenUtil.java:
`P3`

해당 객체는 Bean으로까지 등록할 필요 없이 `public final class`로 선언하여 
Util 객체로서 사용하게끔 하는 것 어떨까요?!!

싱글톤으로 관리되어야 하거나 특정 상태를 보존하고 운영해야 하거나 외부 Bean 객체 주입이 필요한 객체인지 의문이 들어서요!!

ex.
```java
public final class JwtTokenUtil {

  private JwtTokenUtil(){}

  public static String extract(String token) {
    return token.substring(TOKEN_HEADER.length());
  }

  public static String addPrefix(String token) {
    return TOKEN_HEADER + token;
  }
}
```

yummygyudon commented on src/main/resources/application-test.yaml:
```diff
@@ -7,9 +7,17 @@ spring:
   client:
-    url: ${LOCAL_AUTH_URL}
\ No newline at end of file
+    url: ${LOCAL_AUTH_URL}
```
`P1`

엇 `test` 프로파일 설정 yaml을 변경하신 이유가 뭘까요?!!

Spring 기본 테스트 환경에서는 memory H2를 사용하게끔 되어있어서 명시 개념에서 지정을 해놓았는데
저의 경우, Local에서 prostgres를 써서 일관된 테스트 환경 보장이 안될 것 같다는 생각이 듭니다...!!

hyunw9 replied:
넵 이 부분은 다시 돌려놓겠습니다 !  !

yummygyudon reviewed (approved):
우선 몇 가지 부분에 대해서 리뷰 몇개 달아보았습니다!!
제가 하고 싶은 다른 내용을 성은이가 이미 달아놓아서 이모지로 공감 표시했습니다!! ㅋㅋㅋㅋㅋㅋㅋ

꽤 큰 작업이었을텐데 고생많으셨어요😚

hyunw9 committed:
[FEAT] Pem Configuration 가독성 개선

hyunw9 committed:
[FEAT] TokenUtil Bean 해제 및 static import

hyunw9 committed:
[SETTING] test.yaml 복구
